src/sdl/sdl_touch.c

```diff
@@ -0,0 +1,252 @@
+      event.touch.type      = type;
+      event.touch.display   = (ALLEGRO_DISPLAY*)disp;
+      event.touch.timestamp = timestamp;
+      event.touch.id        = id;
+      event.touch.x         = x;
+      event.touch.y         = y;
+      event.touch.dx        = dx;
+      event.touch.dy        = dy;
+      event.touch.primary   = primary;
+
+      _al_event_source_lock(&touch_input->touch_input.es);
+      _al_event_source_emit_event(&touch_input->touch_input.es, &event);
+      _al_event_source_unlock(&touch_input->touch_input.es);
+   }
+
+   if (touch_input->touch_input.mouse_emulation_mode != ALLEGRO_MOUSE_EMULATION_NONE) {
+      mouse_state.x = (int)x;
+      mouse_state.y = (int)y;
+      if (type == ALLEGRO_EVENT_TOUCH_BEGIN)
+         mouse_state.buttons++;
+      else if (type == ALLEGRO_EVENT_TOUCH_END)
+         mouse_state.buttons--;
+
+      mouse_state.pressure = mouse_state.buttons ? 1.0 : 0.0; /* TODO */
+
+      _al_event_source_lock(&touch_input->touch_input.mouse_emulation_es);
+      if (want_mouse_emulation_event) {
+
+         switch (type) {
+            case ALLEGRO_EVENT_TOUCH_BEGIN: type = ALLEGRO_EVENT_MOUSE_BUTTON_DOWN; break;
+            case ALLEGRO_EVENT_TOUCH_CANCEL:
+            case ALLEGRO_EVENT_TOUCH_END:   type = ALLEGRO_EVENT_MOUSE_BUTTON_UP;   break;
+            case ALLEGRO_EVENT_TOUCH_MOVE:  type = ALLEGRO_EVENT_MOUSE_AXES;        break;
+         }
+
+         event.mouse.type      = type;
+         event.mouse.timestamp = timestamp;
+         event.mouse.display   = (ALLEGRO_DISPLAY*)disp;
+         event.mouse.x         = (int)x;
+         event.mouse.y         = (int)y;
+         event.mouse.dx        = (int)dx;
+         event.mouse.dy        = (int)dy;
+         event.mouse.dz        = 0;
+         event.mouse.dw        = 0;
+         if (touch_input->touch_input.mouse_emulation_mode != ALLEGRO_MOUSE_EMULATION_5_0_x) {
+            event.mouse.button = 1;
+         }
+         else {
+            event.mouse.button = id;
+         }
+         event.mouse.pressure  = mouse_state.pressure;
+
+         if (touch_input->touch_input.mouse_emulation_mode != ALLEGRO_MOUSE_EMULATION_5_0_x) {
+            al_set_mouse_xy(event.mouse.display, event.mouse.x, event.mouse.y);
+         }
+
+         _al_event_source_emit_event(&touch_input->touch_input.mouse_emulation_es, &event);
+      }
+      _al_event_source_unlock(&touch_input->touch_input.mouse_emulation_es);
+   }
+}
+
+void _al_sdl_touch_input_event(SDL_Event *e)
+{
+   if (!touch_input)
+      return;
+
+   ALLEGRO_EVENT_TYPE type;
+
+   ALLEGRO_DISPLAY *d = NULL;
+   /* Use the first display as event source if we have any displays. */
+   ALLEGRO_SYSTEM *s = al_get_system_driver();
+   if (_al_vector_size(&s->displays) > 0) {
+      void **v = (void **)_al_vector_ref(&s->displays, 0);
+      d = *v;
+   }
+   if (!d) {
+      return;
+   }
+
+   int touchId = e->tfinger.fingerId;
+
+   touch_input->state.touches[touchId].x = e->tfinger.x;
```

Comment:
Shouldn't you multiply-in the display dimensions here for the al_get_touch_state to work correctly?

Reply:
Oh, yeah, likely - haven't tested that.